platform/ios/src/MGLMapView.mm

```diff
@@ -2889,139 +2861,277 @@ - (CGRect)accessibilityFrame
             CLLocationCoordinate2D coordinateA = [[self annotationWithTag:tagA] coordinate];
             CLLocationCoordinate2D coordinateB = [[self annotationWithTag:tagB] coordinate];
             CGPoint pointA = [self convertCoordinate:coordinateA toPointToView:self];
             CGPoint pointB = [self convertCoordinate:coordinateB toPointToView:self];
-            CGFloat deltaA = hypot(pointA.x - centerPoint.x, pointA.y - centerPoint.y);
-            CGFloat deltaB = hypot(pointB.x - centerPoint.x, pointB.y - centerPoint.y);
+            CGFloat deltaA = HYPOT_SQUARED(pointA.x - centerPoint.x, pointA.y - centerPoint.y);
+            CGFloat deltaB = HYPOT_SQUARED(pointB.x - centerPoint.x, pointB.y - centerPoint.y);
             return deltaA < deltaB;
         });
+        MGL_SIGNPOST_END(self.log, signpost, "gae/sort-annotations");
         
-        NSUInteger annotationIndex = index - visibleAnnotationRange.location;
-        MGLAnnotationTag annotationTag = visibleAnnotations[annotationIndex];
-        MGLAssert(annotationTag != MGLAnnotationTagNotFound, @"Can’t get accessibility element for nonexistent or invisible annotation at index %li.", (long)index);
-        return [self accessibilityElementForAnnotationWithTag:annotationTag];
+        // Elements
+        MGL_SIGNPOST_BEGIN(self.log, signpost,  "gae/annotation-elements");
+        NSMutableArray *elements = [NSMutableArray arrayWithCapacity:visibleAnnotations.size()];
+        
+        for (auto it = visibleAnnotations.begin(); it != visibleAnnotations.end(); ++it)
+        {
+            id element = [self accessibilityElementForAnnotationWithTag:*it];
+            [elements addObject:element];
+        }
+        MGL_SIGNPOST_END(self.log, signpost, "gae/annotation-elements", "%lu", (unsigned long)elements.count);
+
+        _sortedVisibleAnnotationElements = elements;
+        _cachedAccessibilityValue = nil;
     }
+    NSUInteger numberOfAnnotations = _sortedVisibleAnnotationElements.count;
     
-    // Visible place features
-    NSArray *visiblePlaceFeatures = self.visiblePlaceFeatures;
-    NSRange visiblePlaceFeatureRange = NSMakeRange(NSMaxRange(visibleAnnotationRange), visiblePlaceFeatures.count);
-    if (NSLocationInRange(index, visiblePlaceFeatureRange))
-    {
-        visiblePlaceFeatures = [visiblePlaceFeatures sortedArrayUsingComparator:^NSComparisonResult(id <MGLFeature> _Nonnull featureA, id <MGLFeature> _Nonnull featureB) {
+    // Places
+    NSArray *sortedPlaceFeatures = nil;
+    NSArray *visiblePlaceFeatures;
+    if (!_sortedVisiblePlaceElements) {
+        
+        MGL_SIGNPOST_BEGIN(self.log, signpost, "gae/query-places");
+        NSArray *placeStyleLayerIdentifiers = [self.style.placeStyleLayers valueForKey:@"identifier"];
+        
+        // TODO:
+        /*NSArray **/visiblePlaceFeatures = [self visibleFeaturesInRect:bounds inStyleLayersWithIdentifiers:[NSSet setWithArray:placeStyleLayerIdentifiers]];
```

Comment:
What’s the to-do item here?

Reply:
I lifted `visiblePlaceFeatures` above the if, so that I could use it further below.